### ISSUE TITLE
MA0040 should not be reported if invoked IAsyncEnumerable method will receive configured cancellation token

### ISSUE THREAD

**Issue by drieseng** [meziantou/Meziantou.Analyzer#615]
When a method returning an **IAsyncEnumerable\<T>** that takes an **CancellationToken** argument annotated with **EnumeratorCancellationAttribute** is invoked without **CancellationToken**, then **MA0040** should not be reported if the awaitable is configured to automatically pass a given **CancellationToken** when iterating.

**Minimal code**

```cs
using System.Runtime.CompilerServices;
using System.Threading;

public static class Program
{
    private static async Task Main()
    {
        var cancellationTokenSource = new CancellationTokenSource();
        cancellationTokenSource.Cancel();

        await foreach (string x in Run("x").WithCancellation(cancellationTokenSource.Token).ConfigureAwait(false))
        {
            Console.WriteLine(x);
        }
    }

    private static async IAsyncEnumerable<string> Run(string? value, [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        if (value is null)
        {
            throw new ArgumentNullException(nameof(value));
        }

        await Task.Delay(1000, cancellationToken).ConfigureAwait(false);

        yield return "X";
        yield return "Y";
    }
}
```

**Expected result**

The **MA0040** diagnostic is not reported as the `Run(...)` will effectively receive the **CancelleationToken** that is configured using the `WithCancellation(...)` extension method.

This is proven by the fact that running this app will immediately throw a **TaskCanceledException**.

**Actual result**

The following diagnostic is reported:

error MA0040: Use an overload with a CancellationToken, available tokens: cancellationTokenSource.Token (https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0040.md)

**Comment by meziantou** [comment]
There are 2 potential fixes:
- Do not report the rule when `WithCancellation` is used on the expression
- Update the fixer to remove `WithCancellation` as it's not needed when the parameter is used

**Comment by drieseng** [comment]
@meziantou I'm ok with either of the proposed fixes. Perhaps a slight preference for the second one, but that's only because that's what the diagnostic is proposing ("Use an overload with a CancellationToken...").
